### PR TITLE
Rework include file/path handling so we can install the compiler

### PR DESCRIPTION
**Pull request by ChrisDodd**
- look for p4include in the same dir as the executable to override the
  install p4include path.
- install includes in the build directory so tests run.

### CODE REVIEW

**Comment by ChrisDodd** [comment]
One major question here is whether the include files should be installed directly in $prefix ( /usr/local by default), or in a package specific directory (/usr/local/share/p4c/p4include)?

File: frontends/common/options.cpp [hunk 1]
```diff
@@ -130,6 +133,38 @@ void CompilerOptions::setInputFile() {
     }
 }
 
+std::vector<const char*>* CompilerOptions::process(int argc, char* const argv[]) {
+    char buffer[PATH_MAX];
+    int len;
+    struct stat st;
+    if ((len = readlink("/proc/self/exe", buffer, sizeof(buffer))) > 0 ||
```
Comment (mbudiu-vmw) [inline frontends/common/options.cpp]: Is this stuff portable on other operating systems?
Maybe it should be in a Linux-specific library. We haven't been too careful about this kind of system dependences so far.

File: frontends/common/options.cpp [hunk 2]
```diff
@@ -130,6 +133,38 @@ void CompilerOptions::setInputFile() {
...
+    } else if (strchr(argv[0], '/')) {
+        getcwd(buffer, sizeof(buffer));
+        strncat(buffer, argv[0], sizeof(buffer) - strlen(buffer) - 1);
+    } else if (getenv("_")) {
+        strncpy(buffer, getenv("_"), sizeof(buffer));
+        buffer[sizeof(buffer) - 1] = 0;
+    } else {
+        buffer[0] = 0; }
+
+    if (char *p = strrchr(buffer, '/')) {
```
Comment (mbudiu-vmw) [inline frontends/common/options.cpp]: It seems that you are assuming that the include path shares the same prefix with the binary path.
Is this a common assumption?

File: frontends/common/options.cpp [hunk 3]
```diff
@@ -145,8 +180,8 @@ FILE* CompilerOptions::preprocess() {
 #else
         std::string cmd("cpp");
 #endif
-        cmd += cstring(" -undef -nostdinc -I") +
-                p4includePath + " " + preprocessor_options + " " + file;
+        cmd += cstring(" -undef -nostdinc -I") + (isv1() ? p4_14includePath : p4includePath)
```
Comment (mbudiu-vmw) [inline frontends/common/options.cpp]: Shouldn't the p4includePath be appended at the end?
This way we cannot override it with -I.

File: Makefile.am [hunk 4]
```diff
@@ -85,6 +87,22 @@ BUILT_SOURCES += \
...
+# FIXME -- should these be in $(pkgdatadir) instead of $(prefix)?  See configure.ac
+# FIXME -- and check-install-headers: below
+p4includedir = $(prefix)/p4include
+p4_14includedir = $(prefix)/p4_14include
+p4include_HEADERS += $(srcdir)/p4include/core.p4 $(srcdir)/p4include/v1model.p4
+
+#### For testing, install headers in the build directory
+
+check-install-headers:
+	@$(MAKE) install-data prefix=.
```
Comment (cc10512) [inline Makefile.am]: should this be ```@$(MAKE) install-data prefix=$(builddir)```?

**Comment by mbudiu-vmw** [comment]
One other major question is whether we need a versioning scheme for included files.

File: frontends/common/options.cpp [hunk 5]
```diff
@@ -130,6 +133,38 @@ void CompilerOptions::setInputFile() {
...
+        (len = readlink("/proc/curproc/exe", buffer, sizeof(buffer))) > 0 ||
+        (len = readlink("/proc/curproc/file", buffer, sizeof(buffer))) > 0 ||
+        (len = readlink("/proc/self/path/a.out", buffer, sizeof(buffer))) > 0) {
+        buffer[len] = 0;
+    } else if (argv[0][0] == '/') {
+        strcpy(buffer, argv[0]);
+    } else if (strchr(argv[0], '/')) {
+        getcwd(buffer, sizeof(buffer));
+        strncat(buffer, argv[0], sizeof(buffer) - strlen(buffer) - 1);
+    } else if (getenv("_")) {
```
Comment (mbudiu-vmw) [inline frontends/common/options.cpp]: I actually don't know what getenv("_") is.
Maybe you can add a comment?

File: frontends/common/options.cpp (hunk 1 above)
Comment (cc10512) [inline frontends/common/options.cpp]: If I understand correctly this code is finding the directory from where the program was invoked. Why not use dirname(argv[0])? It is much more portable than using /proc. This implementation will not work on MacOS.

Comment (ChrisDodd) [inline frontends/common/options.cpp]: Its portable to Linux, *BSD, and Solaris, and it won't fail badly on other platforms.  The fallback using argv[0] will work on OSX and I believe on Windows (cygwin/msys), though I haven't tried the latter

File: frontends/common/options.cpp (hunk 2 above)
Comment (ChrisDodd) [inline frontends/common/options.cpp]: This is for running tests with a binary in the build directory -- iff there's a p4include directory in the same directory as the executable, it overrides the installed config directory.

File: frontends/common/options.cpp (hunk 5 above)
Comment (ChrisDodd) [inline frontends/common/options.cpp]: Some shells (bash, ksh, and zsh) set the `_` environment variable to the full path of the binary they're about to execute.

File: Makefile.am (hunk 4 above)
Comment (ChrisDodd) [inline Makefile.am]: Possibly, though I'm not sure under what circumstances $(builddir) could be anything other than .

File: frontends/common/options.cpp (hunk 1 above)
Comment (ChrisDodd) [inline frontends/common/options.cpp]: dirname(argv[0]) is the fallback and only works if the program was invoked with an absolute path.  We use cwd+dirname(argv[0]) for a relative path.  We could add searching through the $PATH envvar for cases where there is no path in argv[0].

**Comment by ChrisDodd** [comment]
> One other major question is whether we need a versioning scheme for included files.

That's part of why I was thinking $(pkgdatadir) might be better than $(prefix) -- the former contains the package name which can include a version string, thought we don't currently version p4c

File: frontends/common/options.cpp (hunk 3 above)
Comment (ChrisDodd) [inline frontends/common/options.cpp]: I left it unchanged from before, but you're correct, the standard directory should be last in the path (that's the way C compilers do it; gcc has a special -idirafter option to add directories after the standard set)

**Pull request merged**